Xamarin.Forms.Platform.Android/AppCompat/CheckBoxRendererBase.cs

```diff
@@ -162,6 +162,11 @@ protected virtual void OnElementChanged(ElementChangedEventArgs<CheckBox> e)
 
 		protected virtual void OnElementPropertyChanged(object sender, PropertyChangedEventArgs e)
 		{
+			if (this.IsDisposed())
```

Comment:
If this is false shouldn't we unwind the control as if the element has been set to null? 

Shouldn't some or all of the code that's inside dispose be executed if this condition is false?

Reply:
> If this is false shouldn't we unwind the control as if the element has been set to null?

If this is false, we just keep going; the Object is active, not being collected. 

Under most conditions, when we are done with the Object we call `Dispose(true)`, which handles all of the managed unwinding, including unsubscribing from the Element's `ElementPropertyChanged` event. At which point the Element no longer holds a reference to this Object, so we don't have to worry if the Element's property changes. And the garbage collectors are free to do their jobs after that.

Under the specific conditions we're dealing with in this PR, the garbage collectors decided that the Element and the Object could be collected, and started that process. We don't get any notification of this, so there's no opportunity for us to call `Dispose(true)` and handle the unwinding. And we don't really need to - all the referenced objects are going away, so our managed cleanup wouldn't have any effect anyway. The only exception is this circumstance where the Element gets "saved" by a Binding turning a weak reference into a strong one. When that happens, we run into this "partially collected/disposed" scenario, and we have to account for it.

Eventually, the Object _will_ have `Dispose(false)` called (as part of finalization), but there's no cleanup for us to do at that point.

Reply:
Is there a scenario where *IsDisposed* will be true and then become false? 
Where the Handle will be zero and then get restored to something? Is that the weak reference scenario you are talking about?

Reply:
Not in these circumstances, no. Unless someone specifically writes a renderer to work that way, and I can't immediately see a way to do that which makes sense; the Java peer is on the way out by this point, so what are you even rendering? 

By the time we're in this situation, it's a dead renderer walking; it's a zombie. The Element and the Renderer are not connected to the application at all, except for the weak reference from the binding. The weak reference _may_ give them a reprieve which lasts for the duration of the property changed handler. After that, they're lost, like tears in the rain.

Reply:
> it's a dead renderer walking; it's a zombie.

Right. So at the point that the managed and unmanaged connection has been severed shouldn't we aggressively disconnect everything? The renderer is a ticking time bomb once the Handle is zero and anything from the managed side that tries to call into the unmanaged will cause the ODE exception to be raised.

I'm not saying at this point we should call the managed disposed specifically just that the cleanup code inside the dispose should be called.

For example, since the nature of this check is to exit early when the Handle is zero then why don't we unhook from the OnElementPropertyChanged event?

I feel like the second the Handle transitions to zero then all of this code should be executed
```C#
				_tracker?.Dispose();
				_tracker = null;

// Maybe these two lines if they don't cause an ODE and are just removing the event on the managed side
				SetOnCheckedChangeListener(null);
				OnFocusChangeListener = null;

				if (Element != null)
				{
					Element.PropertyChanged -= OnElementPropertyChanged;
				}
```

Reply:
> Right. So at the point that the managed and unmanaged connection has been severed shouldn't we aggressively disconnect everything?

No, because everything is already going away. Once that connection is severed, we're in a sort of Limbo where anything that tries to communicate across the bridge will throw an exception.

> why don't we unhook from the OnElementPropertyChanged event

I suppose we could, but the only thing we'd be preventing is subsequent calls to the property changed handler, which is already guarded. It's only worth doing if there will be subsequent calls; otherwise, we're just adding an extra delegate remove call. 

> I feel like the second the Handle transitions to zero then all of this code should be executed

The second the handle transitions to zero, anything we do in the renderer is at risk of throwing the exception. And none of it is necessary, because all of that stuff will be immediately garbage collected. The only reason we unhook all that stuff during a deterministic dispose is to help the garbage collector.

Reply:
> No, because everything is already going away. Once that connection is severed, we're in a sort of Limbo where anything that tries to communicate across the bridge will throw an exception.

> The second the handle transitions to zero, anything we do in the renderer is at risk of throwing the exception. And none of it is necessary, because all of that stuff will be immediately garbage collected

Yea I guess that's why I'm saying we should sever everything that connects the xplat element to the managed android component thus limiting the possibility of anything reaching across the bridge. One the handle is zero and we're in limbo literally all connections from the renderer to the Element should be considered dangerous

For example all the IsDisposed checks should probably also be added to the tracker as well here yea?

https://github.com/xamarin/Xamarin.Forms/blob/92b1e07266c50798c00ff89e1a50a8d10e65b578/Xamarin.Forms.Platform.Android/VisualElementTracker.cs#L212

Which is where I was going with saying that the tracker should be cleaned up once we detect the handle going to zero. 

Android Handle Goes to Zero => all connections to Element are severed instantly 
And then if possible whatever native events and listeners are removed but that might not be possible once a class is in limbo....

So the check in property changed would be something like
```C#
if (this.IsDisposed())
{
	this.Cleanup(); //this removes property handler and whatever else it can
	_tracker.Cleanup() // this removes any connections the tracker has to element or tracker just has a similar IsDisposed check that calls Cleanup()
	return;
}
```


> The only reason we unhook all that stuff during a deterministic dispose is to help the garbage collector.

Yea which is why in one version of the conversation I'd almost recommend that we just completely get rid of dispose calls and turn them all into CleanUp calls unless it's for something large like images.  All dispose does is save us an extra pass of the GC and I don't feel like enough things get disposed that this is helpful overall and it's more problematic to not just let the GC do it naturally. But I think that's a different discussion, a different PR, and something that needs additional measurements  :-) 


Are there scenarios where deterministic dispose doesn't get called and dispose is only going to be called by the finalizer?

Reply:
> For example all the IsDisposed checks should probably also be added to the tracker as well here yea?

Yes. That is a thing that we should do. Mark that as a requested change.

> Which is where I was going with saying that the tracker should be cleaned up once we detect the handle going to zero.
> And then if possible whatever native events and listeners are removed but that might not be possible once a class is in limbo....

Yeah, we can't touch _any_ native stuff safely at this point. That's all in the hands of the GC. 

And we'd still have to do the IsDisposed check for all of the "property changed handlers", because we can't be sure which PCH will be hit first. So any given PCH is going to have to do 

```
if(IsDisposed()) { ... clean things up ...; return; }
```

Where "clean things up" is "unhook all the managed event handlers", and specifically _isn't_ "touch anything unmanaged" because we can't be sure of the state. 

But at the moment, AFAIK, the _only_ managed event that can still fire in this state is PropertyChanged; unhooking those other handlers is probably just putting on a belt with your suspenders. `if(IsDispose() { return;} ` is a lot less work, both for us and the CPU. 

> Yea which is why in one version of the conversation I'd almost recommend that we just completely get rid of dispose calls and turn them all into CleanUp calls unless it's for something large like images. All dispose does is save us an extra pass of the GC and I don't feel like enough things get disposed that this is helpful overall and it's more problematic to not just let the GC do it naturally. But I think that's a different discussion, a different PR, and something that needs additional measurements :-)

We could probably tune our deterministic disposal to be less aggressive and maybe even provide some more control options (like allowing folks to retain renderers they want to reuse in the future). But that has nothing to do with the cause of this bug. This bug is happening _because_ we haven't called Dispose(true) on the renderer; we've never had an opportunity to do so. We've also never had an opportunity to call _any_ sort of cleanup method - the GC beat us to it. 

> Are there scenarios where deterministic dispose doesn't get called and dispose is only going to be called by the finalizer?

Yes. FragmentContainer comes to mind. Or with embedded native controls, depending on the ManageNativeControlLifetime setting. Probably some others.